ledger/common/hash/hash.go

```diff
@@ -0,0 +1,120 @@
+type Hash [HashLen]byte
+
+// default value and default hash value for a default node
+var defaultLeafHash Hash
+
+// EmptyHash is a hash with all zeroes, used for padding
+var EmptyHash Hash
+
+// tree maximum height
+const TreeMaxHeight = 256
+
+// we are currently supporting paths of a size equal to 32 bytes.
+// I.e. path length from the rootNode of a fully expanded tree to the leaf node is 256.
+// A path of length k is comprised of k+1 vertices. Hence, we need 257 default hashes.
+const defaultHashesNum = TreeMaxHeight + 1
+
+// array to store all default hashes
+var defaultHashes [defaultHashesNum]Hash
+
+func init() {
+
+	hasher := hash.NewSHA3_256()
+	copy(defaultLeafHash[:], hasher.ComputeHash([]byte("default:")))
+
+	// Creates the Default hashes from base to level height
+	defaultHashes[0] = defaultLeafHash
+	for i := 1; i < defaultHashesNum; i++ {
+		defaultHashes[i] = HashInterNode(defaultHashes[i-1], defaultHashes[i-1])
+	}
+}
+
+// GetDefaultHashes returns the default hashes of the SMT.
+//
+// For each tree level N, there is a default hash equal to the chained
+// hashing of the default value N times.
+func GetDefaultHashes() [defaultHashesNum]Hash {
+	return defaultHashes
+}
+
+// GetDefaultHashForHeight returns the default hashes of the SMT at a specified height.
+//
+// For each tree level N, there is a default hash equal to the chained
+// hashing of the default value N times.
+func GetDefaultHashForHeight(height int) Hash {
+	return defaultHashes[height]
+}
+
+// HashLeaf generates hash value for leaf nodes (SHA3-256).
+//
+// path must be a 32 byte slice.
+// note that we don't include the keys here as they are already included in the path
+// TODO: delete this function after refactoring ptrie
+func HashLeaf(path Hash, value []byte) Hash {
+	hasher := new256()
+	return hasher.hash256Plus(path, value) // path is always 256 bits
+}
+
+// HashInterNode generates hash value for intermediate nodes (SHA3-256).
+//
+// hash1 and hash2 must each be a 32 byte slice.
+// TODO: delete this function after refactoring ptrie
+func HashInterNode(hash1 Hash, hash2 Hash) Hash {
+	hasher := new256()
+	return hasher.hash256plus256(hash1, hash2) // hash1 and hash2 are 256 bits
+}
+
+// HashLeafIn generates hash value for leaf nodes (SHA3-256)
+// and stores the result in the result input
+//
+// path must be a 32 byte slice.
+// note that we don't include the keys here as they are already included in the path
+func HashLeafIn(path Hash, value []byte) Hash {
+	hasher := new256()
+	return hasher.hash256Plus(path, value) // path is always 256 bits
+}
+
+// HashInterNodeIn generates hash value for intermediate nodes (SHA3-256)
+// and stores the result in the input array.
+//
+// result slice can be equal to hash1 or hash2.
+// hash1 and hash2 must each be a 32 byte slice.
+func HashInterNodeIn(hash1 Hash, hash2 Hash) Hash {
+	hasher := new256()
+	return hasher.hash256plus256(hash1, hash2) // hash1 and hash2 are 256 bits
+}
```

Comment:
In the previous code, we used to provide the result pointer to avoid allocating new slices. In the new code, the return value (array) is passed by value since it is only 32 bytes, this would just use the stack. 
Writing to a result pointer doesn't seem to make things faster, I also guess we're adding the memory non-locality overhead. 
I've benched this below:
```
BenchmarkHash/LedgerSha3-8                         2542154               468 ns/op
BenchmarkHash/GenericSha3-8                        1256364               963 ns/op
BenchmarkHash/LedgerSha3PointerResult-8            2556220               469 ns/op
```